lib/FunctionIsolation/InvokeIsolatedFunctions.cpp

```diff
@@ -23,78 +26,48 @@
-
-    // TODO: we're not using the following contract even if we should, probably
-    //       due to some error in the contract logic.
-    //       Specifically, adding this contract leads to deduce that we do *not*
-    //       need /:root at beginning of the recompile-isolated step.
-    ContractGroup IsolatedToRoot(Isolated,
-                                 0,
-                                 IsolatedRoot,
-                                 0,
-                                 pipeline::InputPreservation::Preserve);
-
-    return {
-      ContractGroup(Root, 0, IsolatedRoot, 0, InputPreservation::Erase)
-    };
-  }
-
-  void registerPasses(llvm::legacy::PassManager &Manager) {
-    Manager.add(new InvokeIsolatedFunctionsPass());
-  }
-};
-
-static pipeline::RegisterLLVMPass<InvokeIsolatedPipe> Y;
-
-class InvokeIsolatedFunctions {
+class InvokeIsolatedFunctionsImpl {
 private:
-  using FunctionInfo = tuple<const model::Function *, BasicBlock *, Function *>;
-  using FunctionMap = std::map<MetaAddress, FunctionInfo>;
+  using FunctionInfo = tuple<const model::Function *,
+                             BasicBlock *,
+                             const Function *>;
+  using FunctionMap = std::map<model::Function::Key, FunctionInfo>;
 
 private:
   const model::Binary &Binary;
-  Function *RootFunction = nullptr;
-  Module *M = nullptr;
+  Function &RootFunction;
+  Module &RootModule;
   LLVMContext &Context;
-  GeneratedCodeBasicInfo &GCBI;
+  GeneratedCodeBasicInfo GCBI;
   FunctionMap Map;
 
 public:
-  InvokeIsolatedFunctions(const model::Binary &Binary,
-                          Function *RootFunction,
-                          GeneratedCodeBasicInfo &GCBI) :
+  InvokeIsolatedFunctionsImpl(const model::Binary &Binary,
+                              llvm::Module &RootModule,
+                              const llvm::Module &FunctionModule) :
     Binary(Binary),
-    RootFunction(RootFunction),
-    M(RootFunction->getParent()),
-    Context(M->getContext()),
-    GCBI(GCBI) {
+    RootFunction(*RootModule.getFunction("root")),
+    RootModule(RootModule),
+    Context(RootModule.getContext()),
+    GCBI(Binary) {
+
+    GCBI.run(RootModule);
 
     model::CNameBuilder NameBuilder = Binary;
     for (const model::Function &Function : Binary.Functions()) {
-      llvm::Function *F = M->getFunction(NameBuilder.llvmName(Function));
+      auto *F = FunctionModule.getFunction(NameBuilder.llvmName(Function));
       revng_assert(F != nullptr);
-      Map[Function.Entry()] = { &Function, nullptr, F };
+      Map[Function.key()] = { &Function, nullptr, F };
     }
 
-    for (BasicBlock &BB : *RootFunction) {
+    for (BasicBlock &BB : RootFunction) {
       revng_assert(not BB.empty());
 
       MetaAddress JumpTarget = getBasicBlockJumpTarget(&BB);
       auto It = Map.find(JumpTarget);
       if (It != Map.end()) {
+        revng_assert(get<1>(It->second) == nullptr,
+                     "are we ok with overriding basic blocks here?");
```

Comment:
Please do not leave questions in asserts, I find it quite confusing.

State your expectations, not questions.